docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,384 @@
+2. The value of `compare_a` or `compare_b` must be changed depending on the desired duty cycle!
+:::
+
+If we decide to modify the value of `compare_a` or `compare_b`, we have to update the configuration for the PWM.
+
+```rust
+config.compare_a += 100; // modified value of `compare_a`
+pwm.set_config(&config); // set the new configuration for PWM
+```
+
+## Analog-to-Digital Converter (ADC)
+
+Now we know how to represent an analog signal using digital signals. There are plenty of cases in which we need to know how to transform an analog signal into a digital one, for example a temperature reading, or the voice of a person. This means that we need to correctly represent a continuous wave of infinite values to a discrete wave of a finite set of values.
+For this, we need to sample the analog signal periodically, in other words to measure the analog signal at a fixed interval of time. This is done by using an **Analog-to-Digital converter**.
+
+The *sampling rate* is the frequency at which a new sample is read. The higher the sampling rate, the more samples we get, so the more accurate the representation of the signal.
+
+The *resolution* is the number of bits which we can use in order to store the value of the sample. The higher the resolution, the more values we can store, so the more accurate the representation.
+
+:::info
+For example, a resolution of 8 bits means that we can encode the analog signal to one in 256 levels.
+:::
+
+![ADCSampling](images/sampling_values.svg)
+
+### Nyquist-Shannon Sampling Theorem
+
+For an analog signal to be represented without loss of information, the conversion needs to satisfy the following formula:
+
+$$
+sampling_f >= 2 \times max_{f}
+$$
+
+The analog signal needs to be sampled at a frequency greater than twice the *maximum frequency* of the signal.
+In other words, we must sample at least twice per cycle.
+
+![NyquistTheorem](images/nyquist_theorem.png)
+
+### Example of analog sensor
+
+A **photoresistor** (or photocell) is a sensor that measures the intensity of light around it. Its internal resistance varies depending on the light hitting its surface, therefore the more light there is, the lower the resistance will be. 
+
+![Photoresistor](images/photoresistor.png)
+
+### ADC in Embassy-rs
+
+On the RP2040, ADC uses an interrupt called `ADC_IRQ_FIFO` to signal whenever a new sample has been processed. This new sample will be stored inside a FIFO. In the Embassy library, this interrupt is already implemented, so all we need to do is bind it and use it in our ADC variable.
+
+```rust
+bind_interrupts!(struct Irqs {
+    ADC_IRQ_FIFO => InterruptHandler;
+});
+
+// ---- fn main() ----
+
+// Initialize peripherals
+let peripherals = embassy_rp::init(Default::default());
+
+// Create ADC driver
+let mut adc = Adc::new(peripherals.ADC, Irqs, Config::default());
+```
+
+:::warning
```

Comment:
I would keep this warning even though, even if the `use` statement was moved.

Reply:
Kept it.